lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp

```diff
@@ -632,17 +632,29 @@ class BroadcastOpConversionPattern
     // ops that support implicit broadcasting. We expect each Op's verify
     // function to assert their arguments to verify that they can broadcast.
 
+    if (srcOp->getUsers().empty()) {
+      // This broadcast chain has already been replaced.
+      rewriter.eraseOp(srcOp);
+      return success();
+    }
+
     mlir::Value input = srcOp.getOperand(0);
-    mlir::Value result = srcOp.getResult();
 
-    if (srcOp->getUsers().empty()) {
-      return rewriter.notifyMatchFailure(
-          srcOp, "ttir.broadcast op should have at least one use.");
+    mlir::Operation *nextOp = srcOp;
+    while (isa<ttir::BroadcastOp>(*nextOp->getUsers().begin())) {
```

Comment:
what does `mlir::isa<ttir::BroadcastOp>` return if `getUsers()` is empty?

Reply:
I'd imagine segfault

Reply:
I think it can be simplified, correct me if I'm wrong

```
auto nextOp = srcOp;
do  {
assert(...)
nextOp = nextOp->getUsers().begin(); 
} while (isa<ttir::BroadcastOp>(*nextOp) && !nextOp->getUsers().empty());

rewriter.replaceAllOpUsesWith(nextOp, input);
rewriter.eraseOp(srcOp);
```

Reply:
The above approach has a bug that nextOp has already moved past the Op we want to replace here:
`rewriter.replaceAllOpUsesWith(nextOp, input);`

So either we use another variable to keep track of the past value or else we need to search the operands of nextOp to go back to the value to want to replace. This becomes cumbersome or indirect.

Reply:
There is another issue that if the `while()` loop ended because `nextOp->getUsers()` is empty, then the following rewriter functions should not be executed. So need to added an `if()` check after the `do/while `loop ends.

Reply:
>The above approach has a bug that nextOp has already moved past the Op we want to replace here:

I see, to me it makes sense to have `lastBcastOp` var to keep track of it. It's still cleaner in a way that you write a loop with only 1 exit point which is easier to comprehend. 

>There is another issue that if the while() loop ended because nextOp->getUsers() is empty, then the following rewriter functions should not be executed. So need to add an if() check after the do/while loop ends.

It should be a no-op if list of users is empty.

Reply:
> > The above approach has a bug that nextOp has already moved past the Op we want to replace here:
> 
> I see, to me it makes sense to have `lastBcastOp` var to keep track of it. It's still cleaner in a way that you write a loop with only 1 exit point which is easier to comprehend.

As I mentioned below, this will require another exit point after the while loop ends, so the total exit points might be same. Let me share with how it will look like on slack.

> 
> > There is another issue that if the while() loop ended because nextOp->getUsers() is empty, then the following rewriter functions should not be executed. So need to add an if() check after the do/while loop ends.
> 
> It should be a no-op if list of users is empty.

The user list is empty because either 1) The consumer has already been deleted. The algorithm works greedily such that the first broadcast Op will replace the consumer for the entire broadcast chain. 2) We could have a dead op due to another pass. This can happen now or in future, unless we run dead-code-elimination after each pass.